### PR TITLE
Sort products in a taxon

### PR DESCRIPTION
**Pull request by pamil**

### CODE REVIEW

File: spec/Document/ProductDocumentSpec.php [hunk 1]
```diff
@@ -69,9 +69,9 @@ function it_has_main_taxon()
     function it_has_taxons()
     {
         $taxons = new Collection();
-        $this->setTaxons($taxons);
+        $this->setProductTaxons($taxons);
```
Comment (pamil) [inline spec/Document/ProductDocumentSpec.php:72]: Should add another spec after it was refactored.

File: spec/Document/TaxonDocumentSpec.php [hunk 2]
```diff
@@ -27,13 +27,6 @@ function it_has_slug()
         $this->getSlug()->shouldReturn('/mug');
     }
 
-    function it_has_position()
-    {
-        $this->setPosition(1);
```
Comment (pamil) [inline spec/Document/TaxonDocumentSpec.php:32]: Needs to be undone.

File: src/Controller/SearchController.php [hunk 3]
```diff
@@ -85,6 +85,16 @@ public function __invoke(Request $request)
             $request->query->set('locale', $channel->getDefaultLocale()->getCode());
         }
 
+        if (!$request->query->has('sort')) {
+            if (null !== $request->get('taxonCode')) {
+                $request->query->set('sort', ['taxonPositionByCode' => [$request->get('taxonCode') => 'ASC']]);
```
Comment (pamil) [inline src/Controller/SearchController.php:90]: Looking for better ideas though.

File: src/Factory/ProductDocumentFactory.php [hunk 4]
```diff
@@ -70,115 +67,121 @@ public function __construct(
...
     /**
-     * @param ProductInterface $syliusProduct
+     * @param ProductInterface $product
      * @param LocaleInterface $locale
      * @param ChannelInterface $channel
      *
      * @return ProductDocument
      */
-    public function createFromSyliusSimpleProductModel(ProductInterface $syliusProduct, LocaleInterface $locale, ChannelInterface $channel)
+    public function createFromSyliusSimpleProductModel(ProductInterface $product, LocaleInterface $locale, ChannelInterface $channel)
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php:86]: This factory needs a refactor, but not now.

File: src/Filter/Widget/Sort.php [hunk 5]
```diff
@@ -57,6 +57,18 @@ public function modifySearch(Search $search, FilterState $state = null, SearchRe
                     continue;
                 }
 
+                if ('taxonPositionByCode' === $field) {
```
Comment (pamil) [inline src/Filter/Widget/Sort.php:60]: Any other ideas? The naming is not perfect.

File: tests/Factory/ProductDocumentFactoryTest.php [hunk 6]
```diff
@@ -41,17 +42,17 @@ public function it_creates_product_document_from_sylius_product_model()
         $syliusProductAttributeValue->setAttribute($syliusProductAttribute);
         $syliusProductAttributeValue->setValue('red');
 
-        $syliusTaxon = new SyliusTaxon();
+        $syliusTaxon = new Taxon();
```
Comment (pamil) [inline tests/Factory/ProductDocumentFactoryTest.php:45]: This tests are so unreadable :/

File: src/Factory/ProductDocumentFactory.php (hunk 4 above)
Comment (psihius) [reply src/Factory/ProductDocumentFactory.php:86]: $product variable actually exists in the method body - then new ProductDocument is created into $product variable

Comment (pamil) [reply src/Factory/ProductDocumentFactory.php:86]: I've done a little refactoring there to follow the class names, `$product` is a Sylius entity, `$productDocument` is its ES representation, `$productView` is its view representation etc.

Comment (psihius) [reply src/Factory/ProductDocumentFactory.php:86]: Ah, sorry, now I see :)

File: src/Controller/SearchController.php [hunk 7]
```diff
@@ -85,6 +85,16 @@ public function __invoke(Request $request)
             $request->query->set('locale', $channel->getDefaultLocale()->getCode());
         }
 
+        if (!$request->query->has('sort')) {
+            if (null !== $request->get('taxonCode')) {
```
Comment (lchrusciel) [inline src/Controller/SearchController.php:89]: `$request->has('taxonCode')`?

Also, why don't we narrowing the parameters scope to query parameters only?

File: src/Factory/ProductDocumentFactory.php [hunk 8]
```diff
@@ -187,7 +190,7 @@ public function createFromSyliusSimpleProductModel(ProductInterface $syliusProdu
      *
      * @return TaxonDocument
      */
-    private function createTaxonDocumentFromSyliusTaxon(TaxonInterface $taxon, $localeCode = null)
+    private function createTaxonDocument(TaxonInterface $taxon, $localeCode = null): TaxonDocument
```
Comment (lchrusciel) [inline src/Factory/ProductDocumentFactory.php:193]: Is there any reason why we don't have `string` type hint?

**Review by lchrusciel — APPROVED** [review]

File: src/Controller/SearchController.php (hunk 7 above)
Comment (pamil) [reply src/Controller/SearchController.php:89]: There is not `has` method on request. Also, this request allows `taxonCode` to be either set as an attribute (route parameter at products in taxon route) and as query parameter (all products route).

File: src/Factory/ProductDocumentFactory.php (hunk 8 above)
Comment (pamil) [reply src/Factory/ProductDocumentFactory.php:193]: Not really, I'll fix this in an another PR.

File: src/Controller/SearchController.php (hunk 3 above)
Comment (Arminek) [reply src/Controller/SearchController.php:90]: Maybe you can move this to your custom `Sort` widget as a default value.

File: src/Controller/SearchController.php (hunk 7 above)
Comment (lchrusciel) [reply src/Controller/SearchController.php:89]: :(

**Review by Arminek — APPROVED** [review]

File: src/Controller/SearchController.php (hunk 3 above)
Comment (pamil) [reply src/Controller/SearchController.php:90]: I'd like to, but I think there's no access to the request there :/

**Pull request merged**